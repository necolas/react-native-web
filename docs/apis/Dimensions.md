# Dimensions

Note: dimensions may change (e.g due to device rotation) so any rendering logic
or styles that depend on these constants should try to call this function on
every render (rather than caching the value) or listen to the `change` event.

## Methods

static **get**(dimension: string)

Get a dimension (e.g., `"window"` or `"screen"`).

Example: `const { height, width } = Dimensions.get('window')`

static **addEventListener**(type, handler)

Add an event handler. Supported events:

* `change`: Fires when a property within the `Dimensions` object changes. The argument to the event handler is an object with `window` and `screen ` properties whose values are the same as the return values of `Dimensions.get('window')` and `Dimensions.get('screen')`, respectively.


static **removeEventListener**(type, handler)

Remove an event handler.

