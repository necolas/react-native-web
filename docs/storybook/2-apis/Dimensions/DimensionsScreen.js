/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import DimensionsChange from './examples/DimensionsChange';
import UIExplorer, {
  AppText,
  Code,
  Description,
  DocItem,
  Section,
  storiesOf,
  TextList
} from '../../ui-explorer';
import React from 'react';

const DimensionsScreen = () => (
  <UIExplorer title="Dimensions" url="2-apis/Dimensions">
    <Description>
      Note: dimensions may change (e.g., due to device rotation) so any rendering logic or styles
      that depend on these constants should try to call this function on every render, rather than
      caching the value.
    </Description>

    <Section title="Methods">
      <DocItem
        name="static get"
        typeInfo="(dimension: string) => Object"
        description="Get a dimension (e.g., `window` or `screen`)."
        example={{
          code: "const { height, width } = Dimensions.get('window')"
        }}
      />

      <DocItem
        name="static set"
        typeInfo="(dimensions: ?{[key:string]: any}) => void"
        description="This should only be called server-side with an estimate
        for initial dimensions to be used when pre-rendering pages on the
        server."
      />

      <DocItem
        name="static addEventListener"
        typeInfo="(type: string, handler: function) => void"
        description={[
          <AppText>Add an event handler. Supported events:</AppText>,
          <TextList
            items={[
              <AppText>
                <Code>change</Code>: Fires when a property within the <Code>Dimensions</Code> object
                changes. The argument to the event handler is an object with <Code>window</Code> and{' '}
                <Code>screen</Code> properties whose values are the same as the return values of{' '}
                <Code>Dimensions.get('window')</Code> and <Code>Dimensions.get('screen')</Code>,
                respectively.
              </AppText>
            ]}
          />
        ]}
        example={{
          render: () => <DimensionsChange />
        }}
      />

      <DocItem
        name="static removeEventListener"
        typeInfo="(type: string, handler: function) => void"
        description="Remove an event handler."
      />
    </Section>
  </UIExplorer>
);

storiesOf('APIs', module).add('Dimensions', DimensionsScreen);
