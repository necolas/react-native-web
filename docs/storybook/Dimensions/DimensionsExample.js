import { storiesOf } from '@kadira/storybook'
import UIExplorer from '../UIExplorer'
import {
  StyleSheet,
  Dimensions,
  Button,
  View,
  ScrollView,
  Text
} from 'react-native'
import React, { Component } from 'react'

const styles = StyleSheet.create({
  action: {
    marginTop: 8,
    marginBottom: 8
  },
  log: {
    fontSize: 6,
    marginTop: 8,
    marginBottom: 8
  }
})

class ChangeEventExample extends Component {
  state = { listened: false, logs: [] }

  render() {
    const { logs, listened } = this.state
    const buttonTitle = listened ? 'Remove listener' : 'Add listener'
    return (
      <View>
        <View style={styles.action}>
          <Button onPress={this.toggle} title={buttonTitle} />
        </View>
        <View style={styles.action}>
          <Button onPress={this.clearLogs} title="Clear logs" />
        </View>
        <ScrollView>
          {logs.map((log, i) =>
            <Text key={i} style={styles.log}>
              {log}
            </Text>
          )}
        </ScrollView>
      </View>
    )
  }

  toggle = () => {
    const { listened } = this.state
    if (listened) {
      Dimensions.removeEventListener('change', this._handleChange)
      this.log('Removed listener')
    } else {
      Dimensions.addEventListener('change', this._handleChange)
      this.log('Added listener')
    }
    this.setState({ listened: !listened })
  }

  _handleChange = ({ window, screen }) => {
    // this is only for logging
    window = JSON.stringify(window)
    screen = JSON.stringify(screen)
    this.log(`Changed. window=${window}. screen=${screen}`)
  }

  log = msg => {
    this.setState({
      logs: [`${new Date().toTimeString()} - ${msg}`, ...this.state.logs]
    })
  }

  clearLogs = () => this.setState({ logs: [] })
}

const examples = [
  {
    title: 'Listen to `change` event',
    description: '',
    render: () => <ChangeEventExample />
  }
]

storiesOf('APIs', module).add('Dimensions', () =>
  <UIExplorer description="" examples={examples} title="Dimensions" />
)
