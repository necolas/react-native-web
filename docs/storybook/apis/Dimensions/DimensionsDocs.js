/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import { storiesOf } from '@kadira/storybook';
import UIExplorer, { DocItem } from '../../ui-explorer';
import React from 'react';

const sections = [
  {
    title: 'Methods',
    entries: [
      <DocItem
        name="static get"
        typeInfo="(dimension: string) => Object"
        description="Get a dimension (e.g., `window` or `screen`)."
        example={{
          code: "const { height, width } = Dimensions.get('window')"
        }}
      />
    ]
  }
];

storiesOf('APIs', module).add('Dimensions', () =>
  <UIExplorer
    description="Note: dimensions may change (e.g., due to device rotation) so any rendering logic or styles that depend on these constants should try to call this function on every render, rather than caching the value."
    sections={sections}
    title="Dimensions"
  />
);
