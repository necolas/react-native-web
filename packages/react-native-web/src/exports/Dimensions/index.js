/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';
import invariant from 'fbjs/lib/invariant';

export type DisplayMetrics = {|
  fontScale: number,
  height: number,
  scale: number,
  width: number
|};

type DimensionsValue = {|
  window?: DisplayMetrics,
  screen?: DisplayMetrics
|};

type DimensionKey = 'window' | 'screen';

type DimensionEventListenerType = 'change';

const win = canUseDOM
  ? window
  : {
      devicePixelRatio: undefined,
      innerHeight: (undefined: any),
      innerWidth: (undefined: any),
      screen: {
        height: (undefined: any),
        width: (undefined: any)
      }
    };

const dimensions = {};
const listeners = {};

export default class Dimensions {
  static get(dimension: DimensionKey): DisplayMetrics {
    invariant(dimensions[dimension], `No dimension set for key ${dimension}`);
    return dimensions[dimension];
  }

  static set(initialDimensions: ?DimensionsValue): void {
    if (initialDimensions) {
      if (canUseDOM) {
        invariant(false, 'Dimensions cannot be set in the browser');
      } else {
        dimensions.screen = initialDimensions.screen;
        dimensions.window = initialDimensions.window;
      }
    }
  }

  static _update() {
    dimensions.window = {
      fontScale: 1,
      height: win.innerHeight,
      scale: win.devicePixelRatio || 1,
      width: win.innerWidth
    };

    dimensions.screen = {
      fontScale: 1,
      height: win.screen.height,
      scale: win.devicePixelRatio || 1,
      width: win.screen.width
    };

    if (Array.isArray(listeners['change'])) {
      listeners['change'].forEach((handler) => handler(dimensions));
    }
  }

  static addEventListener(
    type: DimensionEventListenerType,
    handler: (DimensionsValue) => void
  ): void {
    listeners[type] = listeners[type] || [];
    listeners[type].push(handler);
  }

  static removeEventListener(
    type: DimensionEventListenerType,
    handler: (DimensionsValue) => void
  ): void {
    if (Array.isArray(listeners[type])) {
      listeners[type] = listeners[type].filter((_handler) => _handler !== handler);
    }
  }
}

Dimensions._update();

if (canUseDOM) {
  window.addEventListener('resize', Dimensions._update, false);
}
