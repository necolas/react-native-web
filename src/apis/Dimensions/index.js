/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * @flow
 */

import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment'
import invariant from 'fbjs/lib/invariant'

const win = ExecutionEnvironment.canUseDOM ? window : { screen: {} }

const dimensions = {
  screen: {
    fontScale: 1,
    get height() { return win.screen.height },
    scale: win.devicePixelRatio || 1,
    get width() { return win.screen.width }
  },
  window: {
    fontScale: 1,
    get height() { return win.innerHeight },
    scale: win.devicePixelRatio || 1,
    get width() { return win.innerWidth }
  }
}

class Dimensions {
  static get(dimension: string): Object {
    invariant(dimensions[dimension], 'No dimension set for key ' + dimension)
    return dimensions[dimension]
  }
}

module.exports = Dimensions
