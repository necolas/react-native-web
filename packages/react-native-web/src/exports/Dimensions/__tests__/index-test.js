/* eslint-env jasmine, jest */

import Dimensions from '..';

describe('apis/Dimensions', () => {
  test('get', () => {
    expect(Dimensions.get('screen')).toMatchSnapshot();
    expect(Dimensions.get('window')).toMatchSnapshot();
  });

  test('set', () => {
    expect(() => Dimensions.set({})).toThrow();
  });

  test('addEventListener', () => {
    const handler = jest.fn();
    Dimensions.addEventListener('change', handler);
    Dimensions._update();
    expect(handler).toHaveBeenCalledTimes(1);
    expect(handler).toHaveBeenLastCalledWith({
      window: Dimensions.get('window'),
      screen: Dimensions.get('screen')
    });
  });

  test('removeEventListener', () => {
    const handler = jest.fn();
    Dimensions.removeEventListener('change', handler);
    Dimensions._update();
    expect(handler).toHaveBeenCalledTimes(0);
  });
});
