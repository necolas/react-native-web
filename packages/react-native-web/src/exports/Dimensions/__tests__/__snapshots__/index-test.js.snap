// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apis/Dimensions get 1`] = `
Object {
  "fontScale": 1,
  "height": 0,
  "scale": 1,
  "width": 0,
}
`;

exports[`apis/Dimensions get 2`] = `
Object {
  "fontScale": 1,
  "height": 768,
  "scale": 1,
  "width": 1024,
}
`;
